src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,687 @@
+    function setCapManager(address _capManager) external onlyOwner {
+        capManager = _capManager;
+
+        emit CapManagerUpdated(_capManager);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Performance Fee Functions
+    ////////////////////////////////////////////////////
+
+    /// @notice Owner sets the performance fee on increased assets
+    /// @param _fee The performance fee measured in basis points (1/100th of a percent)
+    /// 10,000 = 100% performance fee
+    /// 500 = 5% performance fee
+    /// The max allowed performance fee is 50% (5000)
+    function setFee(uint256 _fee) external onlyOwner {
+        _setFee(_fee);
+    }
+
+    /// @notice Owner sets the account/contract that receives the performance fee
+    function setFeeCollector(address _feeCollector) external onlyOwner {
+        _setFeeCollector(_feeCollector);
+    }
+
+    function _setFee(uint256 _fee) internal {
+        require(_fee <= FEE_SCALE / 2, "ARM: fee too high");
+
+        // Collect any performance fees up to this point using the old fee
+        collectFees();
+
+        fee = SafeCast.toUint16(_fee);
+
+        emit FeeUpdated(_fee);
+    }
+
+    function _setFeeCollector(address _feeCollector) internal {
+        require(_feeCollector != address(0), "ARM: invalid fee collector");
+
+        feeCollector = _feeCollector;
+
+        emit FeeCollectorUpdated(_feeCollector);
+    }
+
+    /// @notice Transfer accrued performance fees to the fee collector
+    /// This requires enough liquidity assets (WETH) in the ARM that are not reserved
+    /// for the withdrawal queue to cover the accrued fees.
+    function collectFees() public returns (uint256 fees) {
+        uint256 newAvailableAssets;
+        // Accrue any performance fees up to this point
+        (fees, newAvailableAssets) = _feesAccrued();
+
+        if (fee == 0) return 0;
+
+        // Check there is enough liquidity assets (WETH) that are not reserved for the withdrawal queue
+        // to cover the fee being collected.
+        _requireLiquidityAvailable(fees);
+        // _requireLiquidityAvailable() is optimized for swaps so will not revert if there are no outstanding withdrawals.
+        // We need to check there is enough liquidity assets to cover the fees being collect from this ARM contract.
+        // We could try the transfer and let it revert if there are not enough assets, but there is no error message with
+        // a failed WETH transfer so we spend the extra gas to check and give a meaningful error message.
+        require(fees <= IERC20(liquidityAsset).balanceOf(address(this)), "ARM: insufficient liquidity");
+
+        IERC20(liquidityAsset).transfer(feeCollector, fees);
+
+        // Save the new available assets back to storage less the collected fees.
+        lastAvailableAssets = SafeCast.toInt128(SafeCast.toInt256(newAvailableAssets) - SafeCast.toInt256(fees));
```

Comment:
I'm looking for a case where lastAvailableAssets can be negative, and I didn't find one. So I've tried to check if lastAvailableAssets can be negative or not.

Because LP is in an only model when updated in `deposit` and `requestRedeem`, `lastAvailableAssets` cannot be negative. So, the only way for it to become negative is when collecting fees.

To do so, we need to have `newAvailableAssets < fees`. Is it possible?
Thanks to the `_requireLiquidityAvailable` we are sure that 
- `fees < weth.balanceOf(ARM) - outstandingWithdrawals`
while we know that: 
- `newAvailableAssets  == weth.balanceOf(ARM) + _externalWithdrawQueue + steth*price - outstandingWithdrawals > weth.balanceOf(ARM) - outstandingWithdrawals > fees`.

I'm not 100% sure of the reasoning here, but could we use a uint256 instead of int256 for `lastAvailableAssets` then?

Reply:
I did work out the math on this last time, and yes it could get negative before. 

I'm not sure what the status is after the latest round of changes. Will have to check again.

Reply:
Here's how it could be negative.

- Imagine one LP in the pool.
- They deposit 100 ETH. lastAvailableAssets = 100
- 10 ETH yield is earned. Actual Assets = 110, lastAvailableAssets has not changed yet, still 100
- They withdraw 102 ETH. lastAvailableAssets is now -2

Reply:
Ok, I see it now, thanks. My first assumption was wrong!